the-collab-lab/tcl-63-smart-shopping-list
# Create new shopping list

## Description

This code adds new features and improves previous implementations as follows:

### App-wide new dependency:
- **@the-collab-lab/shopping-list-utils** is now added to the dependencies which enables further functionalities. We're using its `generateToken()` for token generation. Please, refer to [@the-collab-lab/shopping-list-utils - npm](https://www.npmjs.com/package/@the-collab-lab/shopping-list-utils) for more information on this package.

### Improvements:
- The `List` component we are now displaying the generated token for users’ future reference.
- If the list is empty, the user gets feedback, letting them know that their list does not contain any items. This improves the users’ experience and gives them a hint to add items.

### New app-wide feature: 
- A `Button` component that can be used throughout the development process. We can style this button according to our styling criteria as the project progresses.
- This newly added component accepts value, label, and event listener props.

### File Modifications and Creations:
- **Modified:** Home.jsx, List.jsx, App.jsx, firebase.js
- **Created:** Button.jsx

### App.jsx updates:
**1.** We are no longer using the test mode `my test list` token for `useStateWithStorage()` custom hook. The token value is initiated as `null` until the user enters their assigned token or generates a new token from the Home page. Please refer to **Home.jsx** implementation further below for more info on this update.

**2.** We Implemented routing decisions based on the value of the token in the `listToken` variable:  

 a. If there is a corresponding value saved in the `localStorage`, the code navigates to the corresponding list assigned to that token.
b. If the user navigates to the List route while the token is still `null`, we are now redirecting them back to the `Home` route by checking the `listToken` value.

**3.** We chose to use the `Navigate` component from `react-router-dom` for redirection purposes, as opposed to the `Link` component. The reason is that `Navigate` provides a more direct way to change routes based on specific conditions or logic, making it a more fitting choice for our application's needs.

### Key functional additions In the Home.jsx:
**1.** The Homepage wireframe *(find in `/_resources/wireframes`)*  was our reference to create the UI of our `Home` component.
**2.** We have used the Button component where we needed to have buttons. Please refer to the [New app-wide feature](#new-feature) section for more information on implementation.
**3.** In order to transfer the value of the user-generated token to our routing stream in the `App` component we are using the `setListToken` function, which is passed on to the Home component from the `App` via props.
**4.** Utilizing the `useState` hook we are creating a two-way binding between the token value which is either entered by the user via the form input or the `generateToken()` function.
**5.** We are utilizing the `generateToken()` function from the `@the-collab-lab/shopping-list-utils` package to generate new tokens, triggered via `createNewList` function.
**6.** Currently we are not creating a new collection entry in the database for each new generated token. Our rationale for this decision was the edge case scenario in which a user keeps generating new tokens without entering any items into the new list. We decided that it is best to only save the newly generated token in the local storage and defer the creation of a new collection entry in the Firebase database to when the user starts adding new items to their shopping list. 

However, we have set up the function for creating a new collection in case we decide to proceed with this alternative later on. For details, refer to the `createCollection` function in `src/api/firebase.js`.

**7.** In order to inform the user of a newly generated token, we are displaying the generated token in the form input field via the two-way binding provided by the `useState` hook.
**8.** We are handling the user input via the `handleTokenInput` function in order to keep the user at an appropriate interval for entering the full token before we submit the form input entry.
**9.** `joinExistentList` function: Allows the user to join an existing shopping list by using an active token. Alternatively, if the user generates a new token, this new token will be displayed in the input field and passed on to this function for submission into the local storage for further processing and routing.



## Related Issue



closes #3 

## Acceptance Criteria




If a user **doesn’t** already have a token:
- [x] They are redirected from the List (`/list`) page back to the home (`/`) page.

If a user **does** already have a token:
- [x] They are automatically redirected to the List (`/list`) page.

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
| ✓    | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

### Before


![Pages before](https://github.com/the-collab-lab/tcl-63-smart-shopping-list/assets/54380958/3d66af65-0f02-4a0d-95ec-53b726321320)

### After



![After](https://github.com/the-collab-lab/tcl-63-smart-shopping-list/assets/54380958/73f5c39d-b840-41a1-9bb3-c0f1765bd836)


## Testing Steps / QA Criteria



- To switch to our branch, open your terminal and enter: `git checkout at-rp-NewShoppingList`
- Start the app by running `npm start`
- Click the **Create New List** button to generate a new token. This will lead you to the `/list` page.
- Since your list is initially empty, you should see a message stating, "Your shopping list is empty. Click on "Add item" to begin!". 

### Note:
- If you refresh the page, you should land on the `/list` page since your token has been saved to `localStorage`. However, trying this in incognito mode should only display the homepage and not the `/list` page.
